jormungandr/src/fragment/pool.rs

```diff
@@ -163,16 +163,154 @@ fn is_transaction_valid<E>(tx: &Transaction<E>) -> bool {
 
 pub(super) mod internal {
     use super::*;
-    use lru::LruCache;
+
+    use std::{
+        collections::HashMap,
+        hash::{Hash, Hasher},
+        ptr,
+    };
+
+    /// Doubly-linked queue with the possibility to remove elements from the middle of the list by
+    /// their keys.
+    struct IndexedDeqeue<K, V> {
+        head: *mut IndexedDequeueEntry<K, V>,
+        tail: *mut IndexedDequeueEntry<K, V>,
+
+        index: HashMap<IndexedDequeueKeyRef<K>, Box<IndexedDequeueEntry<K, V>>>,
+    }
+
+    struct IndexedDequeueEntry<K, V> {
+        key: K,
+        value: V,
+
+        prev: *mut IndexedDequeueEntry<K, V>,
+        next: *mut IndexedDequeueEntry<K, V>,
+    }
+
+    /// A wrapper around the pointer to the key of the queue element. This wrapper forwards the
+    /// implementations of `Eq` and `Hash` to `K`. This is required becuase by default the
+    /// implementations of `Eq` and `Hash` from the pointer itself will be used.
+    struct IndexedDequeueKeyRef<K>(*const K);
+
+    impl<K, V> IndexedDeqeue<K, V>
+    where
+        K: Eq + Hash,
+    {
+        fn new() -> Self {
+            Self {
+                head: ptr::null_mut(),
+                tail: ptr::null_mut(),
+
+                index: HashMap::new(),
+            }
+        }
+
+        fn push_front(&mut self, key: K, value: V) {
+            let mut entry = Box::new(IndexedDequeueEntry {
+                key,
+                value,
+                prev: ptr::null_mut(),
+                next: self.head,
+            });
+            if let Some(head) = unsafe { self.head.as_mut() } {
+                head.prev = &mut *entry;
+            } else {
+                self.tail = &mut *entry;
+            }
+            self.head = &mut *entry;
+            if self
+                .index
+                .insert(IndexedDequeueKeyRef(&entry.key), entry)
+                .is_some()
+            {
+                panic!("inserted an already existing key");
+            }
+        }
+
+        fn push_back(&mut self, key: K, value: V) {
+            let mut entry = Box::new(IndexedDequeueEntry {
+                key,
+                value,
+                prev: self.tail,
+                next: ptr::null_mut(),
+            });
+            if let Some(tail) = unsafe { self.tail.as_mut() } {
+                tail.next = &mut *entry;
+            } else {
+                self.head = &mut *entry;
+            }
+            self.tail = &mut *entry;
+            if self
+                .index
+                .insert(IndexedDequeueKeyRef(&entry.key), entry)
+                .is_some()
+            {
+                panic!("inserted an already existing key");
+            }
+        }
+
+        fn pop_back(&mut self) -> Option<(K, V)> {
+            let tail = unsafe { self.tail.as_mut() }?;
+            let entry = self.index.remove(&IndexedDequeueKeyRef(&tail.key)).unwrap();
```

Comment:
I'm not sure on this, but aren't `entry` (`Box<IndexedDequeueEntry>`) and `tail` (`mut *`) aliases to the same memory location? If so, that breaks Rust aliasing rules

Reply:
Well, the whole idea was to bypass some of Rust rules to write C-style code with minimal overhead.

Reply:
But taking care of multiple aliases is explicitly mentioned in the doc for `unsafe as_mut()`, I guess this has to do with how the compiler does optimizations.

Reply:
Ah, you got a point, I missed that piece of the documentation. From my understanding, completing all operations involving `tail` before calling `HashMap::remove` and manually dropping it afterwards should do the trick. This should be fine, `HashMap::remove` does not actually access the data.